R/join_keys.R

```diff
@@ -0,0 +1,697 @@
+      )
+    }
+  )
+  res <- lapply(list_of_list_of_join_key_set, function(.x) do.call(join_keys, .x))
+  names(res) <- names(join_keys_obj)
+
+  logger::log_trace("JoinKeys keys split.")
+  return(res)
+}
+
+#' @rdname merge_join_keys
+#' @export
+merge_join_keys <- function(join_keys_obj, new_join_keys) {
+  UseMethod("merge_join_keys", join_keys_obj)
+}
+
+#' @rdname merge_join_keys
+#' @export
+merge_join_keys.default <- function(join_keys_obj, new_join_keys) {
+  merge_join_keys(join_keys(join_keys_obj), new_join_keys)
+}
+
+#' Merging a list (or one) of `JoinKeys` objects into the current `JoinKeys` object
+#'
+#' @rdname merge_join_keys
+#'
+#' @param join_keys_obj (`JoinKeys`) object to merge the new_join_keys.
+#' @param new_join_keys  `list` of `JoinKeys` objects or single `JoinKeys` object
+#'
+#' @return a new `JoinKeys` object with the resulting merge.
+#'
+#' @export
+#'
+#' @examples
+#' jk1 <- join_keys()
+#' jk1["ds1", "ds2"] <- "some_col"
+#'
+#' jk2 <- join_keys()
+#' jk2["ds1", "ds3"] <- "new_col"
+#'
+#' merge_join_keys(jk1, jk2)
+#' merge_join_keys(jk1, list(jk2))
+merge_join_keys.JoinKeys <- function(join_keys_obj, new_join_keys) {
+  assert_join_keys(join_keys_obj)
+
+  if (test_join_keys(new_join_keys)) {
+    new_join_keys <- list(new_join_keys)
+  }
+
+  lapply(new_join_keys, assert_join_keys_alike)
+
+  checkmate::assert_list(new_join_keys, types = c("JoinKeys"), min.len = 1)
+
+  for (el in new_join_keys) {
+    join_keys_obj <- modifyList(join_keys_obj, el)
+  }
+
+  logger::log_trace("JoinKeys keys merged.")
+  return(join_keys_obj)
+}
+
+#' Prints `JoinKeys`.
+#'
+#' @inheritParams base::print
+#' @return the `x` parameter
+#'
+#' @export
+print.JoinKeys <- function(x, ...) {
+  check_ellipsis(...)
+  keys_list <- x
+  my_parents <- parents(keys_list)
+  class(keys_list) <- "list"
+  if (length(keys_list) > 0) {
+    cat(sprintf(
+      "A JoinKeys object containing foreign keys between %s datasets:\n",
+      length(keys_list)
+    ))
+    # Hide parents
+    attr(keys_list, "__parents__") <- NULL
+    print.default(keys_list[sort(names(keys_list))])
+  } else {
+    cat("An empty JoinKeys object.")
+  }
+  invisible(x)
+}
+
+# -----------------------------------------------------------------------------
+#
+#
+# Helpers (non-exported)
+#
+
+#' Internal constructor
+#'
+#' @return an empty `JoinKeys` list
+#'
+#' @keywords internal
+new_join_keys <- function() {
+  structure(
+    list(),
+    class = c("JoinKeys", "list")
+  )
+}
+
+#' Helper function to add a new pair to a `JoinKeys` object
+#'
+#' @param join_keys_obj (`JoinKeys`) Object with existing pairs.
+#' @param join_key_obj (`JoinKeySet`) relationship pair to add.
+#'
+#' @keywords internal
+join_pair <- function(join_keys_obj, join_key_obj) {
+  assert_join_keys(join_keys_obj)
+  checkmate::assert_class(join_key_obj, "JoinKeySet")
+
+  dataset_1 <- dataset_1.JoinKeySet(join_key_obj)
```

Comment:
quite unusual name for a function. I think it could be shorter or avoided at all by `names(join_key_obj)`. I think there is no need to write a function for such short code.

Reply:
:facepalm: I thought there was a conflict with `get_keys`, but it is a function that will get removed soon.

Renamed in e0488d67

- `dataset_1.JoinKeySet` -> `get_dataset_1`
- `dataset_2.JoinKeySet` -> `get_dataset_2`
- `keys.JoinKeySet` -> `get_keys.JoinKeySet`
  - (added a TODO tag on the generic to rename this one once `get_keys` generic is removed)